### PR TITLE
revised truth label process for new windows and known labels

### PR DESCRIPTION
**Pull request by BoltzmannBrain**
@subutai please review, some big changes to the label combining process:
- nab/labeler.py is revised to accommodate the new raw labels (i.e. only start timestamps, not windows) and known labels, and functions are cleaned up a bit
- added the known labels as labels/raw/known_labels.json
- updated the raw human labels to reflect anomaly start times, not windows
- updated labels/combined_labels.json per the above changes

**Note**: I'm not completely satisfied with labeler.py yet and plan to modify it a bit more. But changes are tough because of other dependencies throughout NAB, so I wanted to get your feedback on this PR first :)

### CODE REVIEW

**Comment by BoltzmannBrain** [comment]
fixes issue #79

File: nab/labeler.py [hunk 1]
```diff
@@ -54,13 +54,15 @@ def __init__(self, path, corpus):
 
     self.corpus = corpus
     self.getWindows()
-    self.getLabels()
+    
+    if "raw" not in self.path:
```
Comment (subutai) [inline nab/labeler.py]: We should mention in the comments the special substring "raw" in path.

File: nab/labeler.py [hunk 2]
```diff
@@ -76,31 +78,43 @@ def found(t, data):
...
-      if not all(deltas):
-        raise ValueError("In the label file %s, timestamps are not in "
-                         "chronological order." % self.path)
-      
+
+
+  def validateLabels(self):
+    """
+    This is run at the end of the label combining process (see
+    scripts/compine_labels.py) to validate the resulting ground truth windows.
```
Comment (subutai) [inline nab/labeler.py]: spelling: `combine_labels`.  Also, what is it validating?

File: nab/labeler.py [hunk 3]
```diff
@@ -194,52 +208,62 @@ def write(self, destPath):
...
+    is merged into one timestamp -- the ground truth label.
+    The set of known anomaly labels are added as well.
     
     If verbosity>0, the dictionary passedLabels -- the raw labels that did not
     pass the threshold qualification -- is printed to the console.
     """
-    combinedLabels = {}
-    labelIndices = {}
-    passedLabels = {}
+    def setTruthLabels(dataSet, trueAnomalies):
```
Comment (subutai) [inline nab/labeler.py]: Comment explaining this function. Also, any particular reason to define inside method?

File: nab/labeler.py [hunk 4]
```diff
@@ -194,52 +208,62 @@ def write(self, destPath):
...
-    if len(self.userLabels) == 0:
+    self.nLabelers = len(self.userLabels)
+    if self.nLabelers == 0:
       raise ValueError("No users labels found")
 
-    self.nlabelers = len(self.userLabels)
-
     
-  def combineRawLabels(self):
+  def combineLabels(self):
```
Comment (subutai) [inline nab/labeler.py]: This method is getting pretty long and hard to follow. I would recommend breaking into smaller parts, and documenting each part really clearly.

File: nab/labeler.py [hunk 5]
```diff
@@ -194,52 +208,62 @@ def write(self, destPath):
...
+    Combines raw user labels to create set of true anomaly labels.
+    A buffer is used to bucket labels that identify the same anomaly. The buffer
     is half the estimated window size of an anomaly -- approximates an average
     of two anomalies per dataset, and no window can have >1 anomaly.
-    After merging, a label becomes a true anomaly if it was labeled by a
-    proportion of the users greater than the defined threshold.
+    After bucketing, a label becomes a true anomaly if it was labeled by a
+    proportion of the users greater than the defined threshold. Then the bucket
+    is merged into one timestamp -- the ground truth label.
+    The set of known anomaly labels are added as well.
```
Comment (subutai) [inline nab/labeler.py]: Are the known anomaly labels treated as if every user had entered an anomaly at the same timestamp?

File: scripts/combine_labels.py [hunk 6]
```diff
@@ -101,6 +106,12 @@ def main(args):
                       labelers for a record to be considered anomalous (should \
                       be a number between 0 and 1)")
   
+  parser.add_argument("--windowSize",
```
Comment (subutai) [inline scripts/combine_labels.py]: I think we should make this a constant in the code. This is not something a NAB user would be setting.

File: scripts/combine_labels.py [hunk 7]
```diff
@@ -101,6 +106,12 @@ def main(args):
...
+                      default=0.10,
+                      type=float,
+                      help="Set the size of the standard anomaly window, as a \
```
Comment (subutai) [inline scripts/combine_labels.py]: This is the size of all anomaly windows combined, not the size of each window.

File: nab/labeler.py (hunk 3 above)
Comment (BoltzmannBrain) [inline nab/labeler.py]: Because it's only called in combineLabels() I thought it would be tidier, similar style to found() in getWindows(). No problem to move it outside the method though.

File: nab/labeler.py (hunk 5 above)
Comment (BoltzmannBrain) [inline nab/labeler.py]: They're added as the raw labels are combined, but they skip the bucket-merge combining process. So it is as if the users entered start timestamps for the known data files that would have resulted in them being bucketed together, surpassing the threshold, and merging into what we have set as the ground truth timestamps for these data files. I'll add to the comments to clarify, but this is also explained in the writeup.

File: scripts/combine_labels.py (hunk 7 above)
Comment (BoltzmannBrain) [inline scripts/combine_labels.py]: Okay, this will be removed following your previous comment -- to delete the windowSize user parameter.

File: nab/labeler.py (hunk 5 above)
Comment (subutai) [inline nab/labeler.py]: OK, thanks.

File: nab/labeler.py (hunk 3 above)
Comment (subutai) [inline nab/labeler.py]: I'm fine leaving it here. Just curious.

File: nab/labeler.py (hunk 4 above)
Comment (BoltzmannBrain) [inline nab/labeler.py]: Agreed, I think I'll make separate functions for "bucketing" and "merging" timestamps.

**Comment by BoltzmannBrain** [comment]
@subutai please review.
I added scripts/create_empty_label_file.py to this PR because it should not create entries for the known label files.

**Comment by subutai** [comment]
:+1:
